drgn_tools/lsmod.py

```diff
@@ -0,0 +1,165 @@
+    ):
+        dep_mod = []
+        for depuse in list_for_each_entry(
+            "struct module_use", use.source_list.address_of_(), "source_list"
+        ):
+            dep_mod.append(depuse.source.name.string_().decode("utf-8"))
+        print(
+            hex(use.mkobj.mod.value_()).ljust(5),
+            "\t",
+            use.name.string_().decode("utf-8").ljust(15),
+            "\t",
+            str(use.core_layout.size.value_()).rjust(10),
+            str(use.refcnt.counter.value_() - 1).rjust(6),
+            "\t",
+            ", ".join(dep_mod),
+        )
+
+    # Module details
+    for use in list_for_each_entry(
+        "struct module", prog["modules"].address_of_(), "list"
+    ):
+        print("\n")
+        print("MODULE NAME:".ljust(15), use.name.string_().decode("utf-8"))
+        print("PARAM COUNT:", str(use.num_kp.value_()))
+        print("ADDRESS    :", hex(use.num_kp.address_of_()))
+        print(
+            "PARAMETER".ljust(25),
+            "ADDRESS".rjust(15),
+            "TYPE".rjust(25),
+            "VALUE".rjust(10),
+        )
+        ker_parm = hex(use.kp.value_())
+        for i in range(0, use.num_kp.value_()):
+            parm_valf = ""
+            parm_val = ""
+            kp = Object(prog, "struct kernel_param", address=int(ker_parm, 16))
+            try:
+                param_type = str(
+                    (
+                        (
+                            (identify_address(kp.ops.get).split(":")[1]).split(
+                                "+"
+                            )[0]
+                        ).split("param_")[1]
+                    )
+                )
+                if "charp" in param_type:
+                    param_type = "charp"
+                    parm_val = str(prog.read(kp.arg.value_(), 2).hex())
+                    if parm_val != "0000":
+                        parm_valf = str(int(parm_val, 16))
+                    elif parm_val == "0000":
+                        parm_valf = "NULL"
+                    else:
+                        parm_valf = "UNKNOWN Str"
+                elif "uint" in param_type:
+                    parm_val = prog.read(kp.arg.value_(), 4).hex()
+                    if len(parm_val) > 2:
+                        parm_valf = str(get_parm_valf(prog, parm_val, kp))
+                    else:
+                        parm_valf = str(
+                            int(str(prog.read(kp.arg.value_(), 2).hex()), 16)
+                        )
+                elif "int" in param_type:
+                    parm_val = prog.read(kp.arg.value_(), 2).hex()
+                    if len(parm_val) > 2:
+                        parm_valf = str(get_parm_valf(prog, parm_val, kp))
+                    else:
+                        parm_valf = str(
+                            int(str(prog.read(kp.arg.value_(), 2).hex()), 16)
+                        )
+                    if str(parm_val) == "ffff":
+                        parm_valf = "-1"
```

Comment:
These integer types are defined something like this:
```
236  STANDARD_PARAM_DEF(byte,	unsigned char,		"%hhu", kstrtou8);
237  STANDARD_PARAM_DEF(short,	short,			"%hi",  kstrtos16);
238  STANDARD_PARAM_DEF(ushort,	unsigned short,		"%hu",  kstrtou16);
239  STANDARD_PARAM_DEF(int,		int,			"%i",   kstrtoint);
240  STANDARD_PARAM_DEF(uint,	unsigned int,		"%u",   kstrtouint);
241  STANDARD_PARAM_DEF(long,	long,			"%li",  kstrtol);
242  STANDARD_PARAM_DEF(ulong,	unsigned long,		"%lu",  kstrtoul);
243  STANDARD_PARAM_DEF(ullong,	unsigned long long,	"%llu", kstrtoull);
```

I think the way to handle this is to map these type names to their corresponding C types, and let drgn handle it. You can create the mapping like this (outside the loop):

```
# From STANDARD_PARAM_DEF in kernel/params.c
integer_params = {
    "byte": "unsigned char *",
    "short": "short *",
    "ushort": "unsigned short *",
    ...
}
```

Then, you can use that mapping in this if statement:
```
...
elif param_type in integer_params:
    value = cast(integer_params[param_type], kp.arg).value
```